addons/addon-edit-s3-bucket-policy/packages/edit-s3-bucket-policy/lib/steps/edit-s3-bucket-policy-service.js

```diff
@@ -0,0 +1,119 @@
+ *  You may not use this file except in compliance with the License.
+ *  A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+const _ = require('lodash');
+const Service = require('@aws-ee/base-services-container/lib/service');
+
+const settingKeys = {
+  environmentsBootstrapBucketName: 'environmentsBootstrapBucketName',
+  studyDataBucketName: 'studyDataBucketName',
+  deploymentBucketName: 'deploymentBucketName',
+};
+
+class EditS3BucketPolicyService extends Service {
+  constructor() {
+    super();
+    this.dependency(['s3Service', 'lockService']);
+  }
+
+  async init() {
+    await super.init();
+    // Setup services and SDK clients
+    [this.s3Service, this.lockService] = await this.service(['s3Service', 'lockService']);
+    this.s3Client = this.s3Service.api;
+  }
+
+  async execute() {
+    const environmentsBootstrapBucketName = this.settings.get(settingKeys.environmentsBootstrapBucketName);
+    const studyDataBucketName = this.settings.get(settingKeys.studyDataBucketName);
+    const deploymentBucketName = this.settings.get(settingKeys.deploymentBucketName);
+
+    const dynamicBucketNames = [environmentsBootstrapBucketName, studyDataBucketName, deploymentBucketName];
+
+    await Promise.all(
+      dynamicBucketNames.map(async bucket => {
+        await this.updateS3PoliciesForSecureTransport(bucket);
+      }),
+    );
+  }
+
+  async replaceS3BucketPolicyStatement(s3BucketName, oldStatementSid, newStatementSid, replacePattern) {
+    this.log.info(`Attempting to update bucket policy for ${s3BucketName}`);
+
+    // Perform locked updates to prevent inconsistencies from race conditions
+    const s3LockKey = `s3|bucket-policy|${s3BucketName}`;
+    // Update S3 bucket policy
+    await this.lockService.tryWriteLockAndRun({ id: s3LockKey }, async () => {
+      // Get existing policy
+      const s3Policy = JSON.parse((await this.s3Client.getBucketPolicy({ Bucket: s3BucketName }).promise()).Policy);
+      this.log.info(s3Policy);
+
+      if (_.isEmpty(s3Policy)) {
+        this.log.info(`No policy attached to ${s3BucketName}.`);
+        return;
+      }
+
+      // Get statements
+      const statements = s3Policy.Statement;
+      let oldStatement;
+
+      if (replacePattern === 'HTTPS') {
+        // Get TLS statement and remove from policy object
+        const tlsStatement = oldStatementSid;
+        statements.forEach((statement, index, object) => {
+          if (statement.Sid === tlsStatement) {
+            this.log.info(`Found statement ${tlsStatement}. Removing.`);
+            oldStatement = statement;
+            object.splice(index, 1);
```

Comment:
It isn't generally considered a good practice to remove items inside a list while you are iterating over them. Seems like `statements.map` function would be a better fit here where you can iterate over each statement and replace the old statement with the new one if it matches `tlsStatement`. That way you get a new list which can be used to replace the old one.

Reply:
Got it! Will try this alternative out.

Reply:
Edited to use `.map`